Импортируем необходимые библиотеки

In [34]:
from sklearn import cross_validation, datasets, metrics, neighbors
from sklearn.grid_search import GridSearchCV
from pandas import DataFrame
import numpy as np

from sklearn.cross_validation import cross_val_score
from sklearn.naive_bayes import GaussianNB, BernoulliNB, MultinomialNB

Загружаем датасеты и выводим несколько строчек выборок из них.

In [4]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [17]:
digits_frame = DataFrame(digits.data)

digits_frame.head()

0   1   2   3   4   5   6   7   8   9  ...  54  55  56  57  58  59  60  61  \
0   0   0   5  13   9   1   0   0   0   0 ...   0   0   0   0   6  13  10   0   
1   0   0   0  12  13   5   0   0   0   0 ...   0   0   0   0   0  11  16  10   
2   0   0   0   4  15  12   0   0   0   0 ...   5   0   0   0   0   3  11  16   
3   0   0   7  15  13   1   0   0   0   8 ...   9   0   0   0   7  13  13   9   
4   0   0   0   1  11   0   0   0   0   0 ...   0   0   0   0   0   2  16   4   

   62  63  
0   0   0  
1   0   0  
2   9   0  
3   0   0  
4   0   0  

[5 rows x 64 columns]

In [18]:
breast_cancer_frame = DataFrame(breast_cancer.data)
breast_cancer_frame.columns = breast_cancer.feature_names
breast_cancer_frame.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension           ...             worst radius  \
0                 0.07871           ...                    25.38   
1                 0.05667           ...                    24.99   
2                 0.05999           ...                    23.57   
3                 0.09744           ...                    14.91   
4                 0.05883           ...                    22.54   

   worst texture  worst perimeter  worst area  worst smoothness  \
0          17.33           184.60      2019.0            0.1622   
1          23.41           158.80      1956.0            0.1238   
2          25.53           152.50      1709.0            0.1444   
3          26.50            98.87       567.7            0.2098   
4          16.67           152.20      1575.0            0.1374   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   worst fractal dimension  
0                  0.11890  
1                  0.08902  
2                  0.08758  
3                  0.17300  
4                  0.07678  

[5 rows x 30 columns]

In [31]:
def bayes_train(data, target):
    average_scores = []
    
    model = BernoulliNB()
    scores = cross_val_score(model, data, target)
    average_scores.append([scores.mean(), 'Bernoulli'])
    
    model = MultinomialNB()
    scores = cross_val_score(model, data, target)
    average_scores.append([scores.mean(), 'Multinomial'])
    
    model = GaussianNB()
    scores = cross_val_score(model, data, target)
    average_scores.append([scores.mean(), 'Gaussian'])
    
    return average_scores

Т.к. модель GaussianNB предполагает признаки с нормальным распределением, модель MultinomialNB используется для целочисленных признаков, а модель BernoulliNB показывает хорошие результаты для двоичных признаков, то ожидаем наилучшее качество модели MultinomialNB для сета 'digits', и модели GaussianNB для сета 'breast_cancer'.

In [40]:
average_scores = bayes_train(digits.data, digits.target)
print 'For dataset \'digits\': \n', average_scores

average_scores = bayes_train(breast_cancer.data, breast_cancer.target)
print 'For dataset \'breast_cancer\': \n', average_scores

For dataset 'digits': 
[[0.82582365077805819, 'Bernoulli'], [0.87087714897350532, 'Multinomial'], [0.81860038035501381, 'Gaussian']]
For dataset 'breast_cancer': 
[[0.62742040285899936, 'Bernoulli'], [0.89457904019307521, 'Multinomial'], [0.9367492806089297, 'Gaussian']]


Заметим, что ожидания оправдались.

В данной задаче использовалась такая метрика качества классификации, как Accuracy, поэтому под "качеством классификации" при ответе на дальнейшие вопросы понимаем именно ее.

1. Каким получилось максимальное качество классификации на датасете breast_cancer?

Ответ: 0.9367492806089297

2. Каким получилось максимальное качество классификации на датасете digits?

Ответ: 0.87087714897350532

3. Какие утверждения из приведенных ниже верны?

(a) На вещественных признаках лучше всего сработал наивный байесовский класси-
фикатор с распределением Бернулли

(b) На вещественных признаках лучше всего сработал наивный байесовский класси-
фикатор с мультиномиальным распределением

(c) Мультиномиальное распределение лучше показало себя на выборке с целыми неот-
рицательными значениями признаков

(d) На вещественных признаках лучше всего сработало нормальное распределение

Ответ: c, d

Тем не менее, из двух конкретных датасетов байесовский классификатор с мультиномиальным распределением сработал численно лучше на выборке с вещественными признаками. Но так как предполагается (как мне кажется), что мы сравниваем качество классификаторов на одном датасете, то будем считать утверждение 'c' верным. 
Еще конкретнее: мультиномиальное распределение показало себя лучше прочих на выборке с целыми неотрицательными значениями признаков, но также среди двух датасетов оказалось лучше для 'breast_cancer' с вещественными признаками.